src/main/java/sample/ClientCredentialsAuthenticationFilter.java

```diff
@@ -0,0 +1,123 @@
+
+        HttpServletRequest req = (HttpServletRequest) request;
+        if (req.getRequestURI().equals("/oauth/token")) {
+            String grantType = req.getParameter("grant_type");
+            if ("client_credentials".equals(grantType)) {
+                String[] credentials = extractCredentials(req);
+                String clientId = credentials[0];
+                String clientSecret = credentials[1];
+                if (!clientId.equals("myclientid") || !clientSecret.equals("myclientsecret")) {
+                    throw new BadCredentialsException("Invalid client credentials");
+                }
+                req.getSession().setAttribute("authentication", true);
+            }
+        }
+
+        chain.doFilter(request, response);
+    }
+
+    private String[] extractCredentials(HttpServletRequest request) {
+        String[] credentials = extractBasicAuthenticationCredentials(request);
+        if (credentials == null) {
+            credentials = extractParameterAuthenticationCredentials(request);
+        }
+        if (credentials == null) {
+            throw new InvalidRequestException("Invalid token request");
+        }
+        return credentials;
+    }
+
+    private String[] extractBasicAuthenticationCredentials(HttpServletRequest request) {
+        String header = request.getHeader("Authorization");
+        if (header != null && header.toLowerCase().startsWith("basic ")) {
+            return extractAndDecodeHeader(header, request);
+        }
+        return null;
+    }
+
+    private String[] extractParameterAuthenticationCredentials(HttpServletRequest request) {
+        String clientId = request.getParameter("client_id");
+        if (clientId != null) {
+            String clientSecret = request.getParameter("client_secret");
+            if (clientSecret == null) {
+                throw new InvalidRequestException("No client_secret parameter provided");
+            }
+            return new String[] { clientId, clientSecret };
+        }
+        return null;
+    }
+
+    // Taken from BasicAuthenticationFilter (spring-security-web)
+    private String[] extractAndDecodeHeader(String header, HttpServletRequest request) {
+
+        byte[] base64Token = header.substring(6).getBytes(UTF_8);
+        byte[] decoded;
+        try {
+            decoded = Base64.getDecoder().decode(base64Token);
+        }
+        catch (IllegalArgumentException e) {
+            throw new BadCredentialsException("Failed to decode basic authentication token");
+        }
+
+        String token = new String(decoded, getCredentialsCharset(request));
+
+        int delim = token.indexOf(":");
+
+        if (delim == -1) {
+            throw new BadCredentialsException("Invalid basic authentication token");
+        }
+        return new String[] { token.substring(0, delim), token.substring(delim + 1) };
+    }
```

Comment:
You got the same problem here as I mentioned in this pull request:
https://github.com/spring-projects-experimental/spring-authorization-server/pull/63